Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,448 @@
+/-
+Copyright (c) 2023 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Damiano Testa
+-/
+import Mathlib.Algebra.Group.Basic
+
+/-!
+
+#  `move_add` a tactic for moving summands in expressions
+
+The tactic `move_add` rearranges summands in expressions.
+
+The tactic takes as input a list of terms, each one optionally preceded by `←`.
+A term preceded by `←` gets moved to the left, while a term without `←` gets moved to the right.
+
+* Singleton input: `move_add [a]` and `move_add [← a]`
+
+  If `⊢ b + a + c` is (a summand in) the goal, then
+  * `move_add [← a]` changes the goal to `a + b + c` (effectively, `a` moved to the left).
+  * `move_add [a]` changes the goal to `b + c + a` (effectively, `a` moved to the right);
+
+  The tactic reorders *all* sub-expressions of the target at the same same.
+  For instance, if `⊢ 0 < if b + a < b + a + c then a + b else b + a` is the goal, then
+  * `move_add [a]` changes the goal to `0 < if b + a < b + c + a then b + a else b + a`
+    (`a` moved to the right in three sums);
+  * `move_add [← a]` changes the goal to `0 < if a + b < a + b + c then a + b else a + b`
+    (`a` again moved to the left in three sums).
+
+* Longer iputs: `move_add [..., a, ..., ← b, ...]`
+
+  If the list contains more than one term, the tactic effectively tries to move each term preceded
+  by `←` to the left, each term not preceded by `←` to the right
+  *maintaining the relative order in the call*.
+  Thus, applying `move_add [a, b, c, ← d, ← e]` returns summands of the form
+  `d + e + [...] + a + b + c`, i.e. `d` and `e` have the same relative position in the input list
+  and in the final rearrangement (and similarly for `a, b, c`).
+  In particular, `move_add [a, b]` likely has the same effect as
+  `move_add [a]; move_add [b]`: first, we move `a` to the right, then we move `b` also to the
+  right, *after* `a`.
+  However, `move_add [← a, ← b]` likely has the same effect as
+  `move_add [b]; move_add [a]`: first, we move `b` to the left, then we move `a` also to the
+  left, *before* `a`.
+  Also note, though, that `move_add [a, b]` may differ `move_add [b]; move_add [a]`,
+  for instance when `a` and `b` are `DefEq`.
```

Comment:
When reading this paragraph I found myself wondering exactly when "likely has the same effect" applies. Am I right in understanding: `move_add [a, b]` and `move_add [a]; move_add [b]` have the same effect if (and only if?) `a` and `b` do not unify? If so, I would start this paragraph by explaining this assumption.

Reply:
I see, thanks! I suppose we could then write something like "if the terms matched by `a` and `b` do not overlap".

Reply:
In any case, is it supposed to be an important use of `move_add` to support overlap, or can we simply say that behaviour in case of overlap is unspecified?